File: app/src/main/java/dev/ragnarok/fenrir/settings/SettingsImpl.java
package dev.ragnarok.fenrir.settings;

import static dev.ragnarok.fenrir.util.Objects.isNull;

import android.content.Context;

public class SettingsImpl implements ISettings {

    private static volatile SettingsImpl instance;
    private final IRecentChats recentChats;
    private final IDrawerSettings drawerSettings;
    private final ISideDrawerSettings sideDrawerSettings;
    private final IPushSettings pushSettings;
    private final ISecuritySettings securitySettings;
    private final IUISettings iuiSettings;
    private final INotificationSettings notificationSettings;
    private final IMainSettings mainSettings;
    private final IAccountsSettings accountsSettings;
    private final IOtherSettings otherSettings;

    private SettingsImpl(Context app) {
        notificationSettings = new NotificationsPrefs(app);
        recentChats = new RecentChatsSettings(app);
        drawerSettings = new DrawerSettings(app);
        sideDrawerSettings = new SideDrawerSettings(app);
        pushSettings = new PushSettings(app);
        securitySettings = new SecuritySettings(app);
        iuiSettings = new UISettings(app);
        mainSettings = new MainSettings(app);
        accountsSettings = new AccountsSettings(app);
        otherSettings = new OtherSettings(app);
    }

    public static SettingsImpl getInstance(Context context) {
        if (isNull(instance)) {
            synchronized (SettingsImpl.class) {
                if (isNull(instance)) {
                    instance = new SettingsImpl(context.getApplicationContext());
                }
            }
        }

        return instance;
    }

    @Override
    public IRecentChats recentChats() {
        return recentChats;
    }

    @Override
    public IDrawerSettings drawerSettings() {
        return drawerSettings;
    }

    @Override
    public ISideDrawerSettings sideDrawerSettings() {
        return sideDrawerSettings;
    }

    @Override
    public IPushSettings pushSettings() {
        return pushSettings;
    }

    @Override
    public ISecuritySettings security() {
        return securitySettings;
    }

    @Override
    public IUISettings ui() {
        return iuiSettings;
    }

    @Override
    public INotificationSettings notifications() {
        return notificationSettings;
    }

    @Override
    public IMainSettings main() {
        return mainSettings;
    }

    @Override
    public IAccountsSettings accounts() {
        return accountsSettings;
    }

    @Override
    public IOtherSettings other() {
        return otherSettings;
    }
}

File: app/src/main/java/dev/ragnarok/fenrir/settings/SettingsImpl.kt
package dev.ragnarok.fenrir.settings

import android.content.Context
import dev.ragnarok.fenrir.settings.ISettings.*

class SettingsImpl private constructor(app: Context) : ISettings {
    private val recentChats: IRecentChats
    private val drawerSettings: IDrawerSettings
    private val sideDrawerSettings: ISideDrawerSettings
    private val pushSettings: IPushSettings
    private val securitySettings: ISecuritySettings
    private val iuiSettings: IUISettings
    private val notificationSettings: INotificationSettings
    private val mainSettings: IMainSettings
    private val accountsSettings: IAccountsSettings
    private val otherSettings: IOtherSettings
    override fun recentChats(): IRecentChats {
        return recentChats
    }

    override fun drawerSettings(): IDrawerSettings {
        return drawerSettings
    }

    override fun sideDrawerSettings(): ISideDrawerSettings {
        return sideDrawerSettings
    }

    override fun pushSettings(): IPushSettings {
        return pushSettings
    }

    override fun security(): ISecuritySettings {
        return securitySettings
    }

    override fun ui(): IUISettings {
        return iuiSettings
    }

    override fun notifications(): INotificationSettings {
        return notificationSettings
    }

    override fun main(): IMainSettings {
        return mainSettings
    }

    override fun accounts(): IAccountsSettings {
        return accountsSettings
    }

    override fun other(): IOtherSettings {
        return otherSettings
    }

    companion object {
        @Volatile
        private var instance: SettingsImpl? = null
        fun getInstance(context: Context): SettingsImpl {
            if (instance == null) {
                synchronized(SettingsImpl::class.java) {
                    if (instance == null) {
                        instance = SettingsImpl(context.applicationContext)
                    }
                }
            }
            return instance!!
        }
    }

    init {
        notificationSettings = NotificationsPrefs(app)
        recentChats = RecentChatsSettings(app)
        drawerSettings = DrawerSettings(app)
        sideDrawerSettings = SideDrawerSettings(app)
        pushSettings = PushSettings(app)
        securitySettings = SecuritySettings(app)
        iuiSettings = UISettings(app)
        mainSettings = MainSettings(app)
        accountsSettings = AccountsSettings(app)
        otherSettings = OtherSettings(app)
    }
}